### PR TITLE
HPCC-27229 Add paging support to FluentUI Grid

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <GordonJSmith@gmail.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [x] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/src/src-react/components/Workunits.tsx [hunk 1]
```diff
@@ -168,10 +174,10 @@ export const Workunits: React.FunctionComponent<WorkunitsProps> = ({
             key: "open", text: nlsHPCC.Open, disabled: !uiState.hasSelection, iconProps: { iconName: "WindowEdit" },
             onClick: () => {
                 if (selection.length === 1) {
-                    window.location.href = `#/workunits/${selection[0].Wuid}`;
+                    window.location.href = `# / workunits / ${selection[0].Wuid} `;
                 } else {
                     for (let i = selection.length - 1; i >= 0; --i) {
-                        window.open(`#/workunits/${selection[i].Wuid}`, "_blank");
+                        window.open(`# / workunits / ${selection[i].Wuid} `, "_blank");
```
Comment (jeclrsg) [inline esp/src/src-react/components/Workunits.tsx:180]: And here

File: esp/src/src-react/hooks/grid.tsx [hunk 2]
```diff
@@ -25,7 +28,7 @@ export function useGrid({ store, query = {}, sort = [], columns, getSelected, fi
     const [grid, setGrid] = React.useState<any>(undefined);
     const [selection, setSelection] = React.useState([]);
 
-    const Grid = React.useMemo(() => () => <DojoGrid
+    const Grid = React.useCallback(() => <DojoGrid
```
Comment (jeclrsg) [inline esp/src/src-react/hooks/grid.tsx:31]: Why switch to useCallback?

File: esp/src/src-react/hooks/grid.tsx [hunk 3]
```diff
@@ -174,31 +189,156 @@ export function useFluentGrid({ data, primaryID, alphaNumColumns = {}, query = {
...
+
+    const copyButtons = React.useMemo((): ICommandBarItemProps[] => [
+        ...createCopyDownloadSelection(constColumns, selection, `${filename}.csv`)
+    ], [constColumns, filename, selection]);
+
+    return [Grid, selection, copyButtons];
+}
+
+export interface useFluentPagedGridProps {
+    store: ESPRequest.Store,
```
Comment (jeclrsg) [inline esp/src/src-react/hooks/grid.tsx:213]: I was wondering about this when I was looking through the consumers of this. The useFluentGrid hook takes in an array of rows, where this one takes a store. Would there be any benefit to having them use the same kind of handle to data?

File: esp/src/src-react/hooks/grid.tsx [hunk 4]
```diff
@@ -174,31 +189,156 @@ export function useFluentGrid({ data, primaryID, alphaNumColumns = {}, query = {
...
+    />, [fluentColumns, items, onColumnClick, renderDetailsHeader, selectionHandler, width]);
 
-    />, [fluentColumns, items, onColumnClick, renderDetailsHeader, renderItemColumn, selectionHandler]);
+    const GridPagination = React.useCallback(() => {
+        return <Stack horizontal horizontalAlign="space-between">
+            <Stack.Item>
+            </Stack.Item>
+            <Stack.Item>
+                <Pagination selectedPageIndex={page} itemsPerPage={pageSize} totalItemCount={total} pageCount={Math.abs(total / pageSize)} format="buttons" onPageChange={index => {
+                    console.log(index);
```
Comment (jeclrsg) [inline esp/src/src-react/hooks/grid.tsx:320]: a console.log here

File: esp/src/package.json [hunk 5]
```diff
@@ -30,57 +30,58 @@
...
     "es6-promise": "^4.2.8",
     "font-awesome": "^4.7.0",
     "formik": "^2.2.9",
-    "query-string": "^7.1.0",
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2",
-    "react-hook-form": "^7.24.1",
-    "react-reflex": "^4.0.3",
+    "query-string": "^7.1.1",
+    "react": "^16.14.0",
```
Comment (jeclrsg) [inline esp/src/package.json:68]: Did this cause problems when it was bumped to 17 in https://github.com/hpcc-systems/HPCC-Platform/pull/15637? I hadn't noticed any.

**Review by jeclrsg — CHANGES_REQUESTED** [review]
@GordonSmith Looks pretty good to me. Just a couple of small things to change. And I had a couple or three questions.

File: esp/src/src-react/hooks/grid.tsx (hunk 2 above)
Comment (GordonSmith) [reply esp/src/src-react/hooks/grid.tsx:31]: Just for consistency as it is really a callback.

File: esp/src/package.json (hunk 5 above)
Comment (GordonSmith) [reply esp/src/package.json:68]: @fluent issues a sibling warning.

File: esp/src/src-react/hooks/grid.tsx [hunk 6]
```diff
@@ -174,31 +189,156 @@ export function useFluentGrid({ data, primaryID, alphaNumColumns = {}, query = {
...
+        styles={{ root: { width, height, maxHeight: height } }}
+    />, [fluentColumns, items, onColumnClick, renderDetailsHeader, selectionHandler, width]);
 
-    />, [fluentColumns, items, onColumnClick, renderDetailsHeader, renderItemColumn, selectionHandler]);
+    const GridPagination = React.useCallback(() => {
+        return <Stack horizontal horizontalAlign="space-between">
+            <Stack.Item>
+            </Stack.Item>
+            <Stack.Item>
+                <Pagination selectedPageIndex={page} itemsPerPage={pageSize} totalItemCount={total} pageCount={Math.abs(total / pageSize)} format="buttons" onPageChange={index => {
```
Comment (GordonSmith) [inline esp/src/src-react/hooks/grid.tsx:319]: Should be `Math.ceil` and not `Math.abs`

File: esp/src/src-react/hooks/grid.tsx [hunk 7]
```diff
@@ -174,31 +189,153 @@ export function useFluentGrid({ data, primaryID, alphaNumColumns = {}, query = {
...
+                <Label htmlFor={"pageSize"}>{nlsHPCC.PageSize}</Label>
+                <Dropdown id="pageSize" options={[
+                    { key: 10, text: "10" },
+                    { key: 25, text: "25" },
+                    { key: 50, text: "50" },
+                    { key: 100, text: "100" },
+                    { key: 250, text: "250" },
+                    { key: 500, text: "500" },
+                    { key: 1000, text: "1000" }
+                ]} defaultSelectedKey={pageSize} onChange={(evt, option) => setPageSize(option.key as number)} />
```
Comment (GordonSmith) [inline esp/src/src-react/hooks/grid.tsx:331]: Needs to recalculate an appropriate pageNum

**Comment by GordonSmith** [comment]
@jeclrsg - round 2, ready for review.

File: esp/src/src-react/components/DFUWorkunits.tsx [hunk 8]
```diff
@@ -60,10 +60,10 @@ export const DFUWorkunits: React.FunctionComponent<DFUWorkunitsProps> = ({
     const [uiState, setUIState] = React.useState({ ...defaultUIState });
 
     //  Grid ---
-    const [Grid, selection, refreshTable, copyButtons] = useGrid({
+    const { Grid, selection, refreshTable, copyButtons } = useGrid({
         store: store || ESPDFUWorkunit.CreateWUQueryStore({}),
         query: formatQuery(filter),
-        sort: [{ attribute: "Wuid", "descending": true }],
+        sort: { attribute: "Wuid", descending: false },
```
Comment (GordonSmith) [inline esp/src/src-react/components/DFUWorkunits.tsx:66]: Should this be true?

File: esp/src/src-react/components/Workunits.tsx [hunk 9]
```diff
@@ -242,10 +252,25 @@ export const Workunits: React.FunctionComponent<WorkunitsProps> = ({
         header={<CommandBar items={buttons} farItems={copyButtons} />}
         main={
             <>
-                <Grid />
+                <SizeMe monitorHeight>{({ size }) =>
+                    <div style={{ width: "100%", height: "100%" }}>
+                        <div style={{ position: "absolute", width: "100%", height: `${size.height}px`, minHeight: `${size.height}px`, maxHeight: `${size.height}px` }}>
```
Comment (GordonSmith) [inline esp/src/src-react/components/Workunits.tsx:257]: min/max probably not needed now

File: esp/src/src-react/components/Workunits.tsx [hunk 10]
```diff
@@ -242,10 +252,25 @@ export const Workunits: React.FunctionComponent<WorkunitsProps> = ({
...
     />;
 };
+
+/*  
+<div style={{ position: "relative", width: "100%", height: "100%" }}>
+<div style={{ position: "absolute", width: "100%", height: `${size.height}px`, minHeight: `${size.height}px`, maxHeight: `${size.height}px` }}>
+    <Grid height={`${size.height}px`} />
+</div>
+</div> 
+*/
```
Comment (GordonSmith) [inline esp/src/src-react/components/Workunits.tsx:276]: Commented code

File: esp/src/src-react/layouts/HolyGrail.tsx [hunk 11]
```diff
@@ -1,28 +1,39 @@
...
         <div style={{ flex: "1 1", display: "flex", minWidth: 0 }} >
-            <div style={{ flex: "0 2" }}>{left}</div>
-            <div style={{ flex: "1 1 auto", minWidth: 1, minHeight: 1 }}>{main}</div>
-            <div style={{ flex: "0 2" }}>{right}</div>
+            <div style={{ flex: "0 2", ...style?.left }}>{left}</div>
+            <div style={{ flex: "1 1 auto", minWidth: 1, minHeight: 1, ...style?.content }}>{main}</div>
+            <div style={{ flex: "0 2", ...style?.right }}>{right}</div>
         </div>
-        <footer style={{ flex: "0 0", minWidth: 0 }}>{footer}</footer>
+        <footer style={{ flex: "0 0", minWidth: 0, ...style?.footer }}>{footer}</footer>
```
Comment (GordonSmith) [inline esp/src/src-react/layouts/HolyGrail.tsx:37]: Revert these changes.

**Review by jeclrsg — APPROVED** [review]
